kernel/timeout.c

```diff
@@ -127,7 +127,22 @@ void z_add_timeout(struct _timeout *to, _timeout_func_t fn,
 		}
 
 		if (to == first()) {
+#if CONFIG_TIMESLICING
+			/*
+			 * This is not ideal, since it does not
+			 * account the time elapsed since the the
+			 * last announcement, and slice_ticks is based
```

Comment:
The elapsed() predicate in this file does exactly that, the driver is required to provide it.  It's a wrapper around z_clock_elapsed() that also handles the case where a timeout is submitted from inside the handler for another timeout (where we want to use the time the timeout was scheduled as "now" and not the potentially later current time -- long bikeshed about that from a few years back).

Reply:
Have you looked the sample I put in the commit message ? That code is failing without this fix because the drivers are unconditionally overriding the current timeout in the function `z_clock_set_timeout` if I remember correctly. Though if there is a contract with the drivers about it we should fix them.

I honestly don't remember all details of this fix, but the problem exists, this test systematically fails, the first slice runs for a longer period.

Reply:
@andy the problem is that if the next timeout is `current_cpu->slice_ticks ` returned by `next_timeout()` the timer should not being set again, since it is already configured. If we call `z_clock_set_timeout` with this value we lost the time between the last announcement and *now*.

Reply:
Having this comparison here is annoying me, but next_timeout() can not simply return a negative value in this case because negative value can mean `K_FOREVER`, can add a parameter in that function though, but not sure if this would be clear.